File: implementations/go/lexer.go
package rcl

import "unicode"

type lexer struct {
	input  []rune
	pos    int
	line   int
	column int
}

type lexerState struct{ pos, line, column int }

func newLexer(s string) *lexer { return &lexer{input: []rune(s), line: 1, column: 1} }
func (l *lexer) snapshot() lexerState { return lexerState{l.pos, l.line, l.column} }
func (l *lexer) restore(st lexerState) { l.pos, l.line, l.column = st.pos, st.line, st.column }

func (l *lexer) nextToken() (token, error) {
	l.skip()
	if l.eof() { return token{typ: tokEOF, line: l.line, col: l.column}, nil }
	line, col, ch := l.line, l.column, l.cur()
	switch ch {
	case '=':
		l.adv(); return token{typ: tokEqual, value: "=", line: line, col: col}, nil
	case ',':
		l.adv(); return token{typ: tokComma, value: ",", line: line, col: col}, nil
	case '.':
		l.adv(); return token{typ: tokDot, value: ".", line: line, col: col}, nil
	case '[':
		l.adv(); return token{typ: tokLBracket, value: "[", line: line, col: col}, nil
	case ']':
		l.adv(); return token{typ: tokRBracket, value: "]", line: line, col: col}, nil
	case '"':
		return l.readString(line, col)
	}
	if ch == '-' && unicode.IsDigit(l.peek()) || unicode.IsDigit(ch) { return l.readNumber(line, col), nil }
	if unicode.IsLetter(ch) || ch == '_' { return l.readIdent(line, col), nil }
	return token{}, &ParseError{Message: "unexpected character", Line: line, Column: col}
}

func (l *lexer) readString(line, col int) (token, error) {
	l.adv()
	out := []rune{}
	for !l.eof() && l.cur() != '"' {
		if l.cur() == '\\' {
			l.adv()
			if l.eof() { return token{}, &ParseError{Message: "unterminated escape", Line: line, Column: col} }
			switch l.cur() {
			case '"': out = append(out, '"')
			case 'n': out = append(out, '\n')
			case 't': out = append(out, '\t')
			case '\\': out = append(out, '\\')
			default: return token{}, &ParseError{Message: "invalid escape", Line: l.line, Column: l.column}
			}
			l.adv(); continue
		}
		out = append(out, l.cur())
		l.adv()
	}
	if l.eof() { return token{}, &ParseError{Message: "unterminated string", Line: line, Column: col} }
	l.adv()
	return token{typ: tokString, value: string(out), line: line, col: col}, nil
}

func (l *lexer) readNumber(line, col int) token {
	out := []rune{}
	if l.cur() == '-' { out = append(out, '-'); l.adv() }
	for !l.eof() && unicode.IsDigit(l.cur()) { out = append(out, l.cur()); l.adv() }
	if !l.eof() && l.cur() == '.' && unicode.IsDigit(l.peek()) {
		out = append(out, '.')
		l.adv()
		for !l.eof() && unicode.IsDigit(l.cur()) { out = append(out, l.cur()); l.adv() }
	}
	return token{typ: tokNumber, value: string(out), line: line, col: col}
}

func (l *lexer) readIdent(line, col int) token {
	out := []rune{}
	for !l.eof() && (unicode.IsLetter(l.cur()) || unicode.IsDigit(l.cur()) || l.cur() == '_') { out = append(out, l.cur()); l.adv() }
	if string(out) == "do" { return token{typ: tokDo, value: "do", line: line, col: col} }
	if string(out) == "end" { return token{typ: tokEnd, value: "end", line: line, col: col} }
	return token{typ: tokIdentifier, value: string(out), line: line, col: col}
}

func (l *lexer) skip() {
	for !l.eof() {
		if unicode.IsSpace(l.cur()) { l.adv(); continue }
		if l.cur() == '#' { for !l.eof() && l.cur() != '\n' { l.adv() }; continue }
		break
	}
}

func (l *lexer) adv() {
	if l.eof() { return }
	if l.cur() == '\n' { l.line++; l.column = 1 } else { l.column++ }
	l.pos++
}
func (l *lexer) cur() rune { return l.input[l.pos] }
func (l *lexer) peek() rune { if l.pos+1 >= len(l.input) { return 0 }; return l.input[l.pos+1] }
func (l *lexer) eof() bool { return l.pos >= len(l.input) }

File: implementations/rust/src/lexer.rs
use crate::error::ParseError;

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum TokenType { Identifier, String, Number, Equal, Comma, Dot, Do, End, LBracket, RBracket, Eof }

#[derive(Clone, Debug)]
pub struct Token { pub typ: TokenType, pub value: String, pub line: usize, pub col: usize }

#[derive(Clone)]
pub struct Lexer { input: Vec<char>, pos: usize, line: usize, col: usize }

impl Lexer {
    pub fn new(s: &str) -> Self { Self { input: s.chars().collect(), pos: 0, line: 1, col: 1 } }
    pub fn snapshot(&self) -> (usize, usize, usize) { (self.pos, self.line, self.col) }
    pub fn restore(&mut self, st: (usize, usize, usize)) { (self.pos, self.line, self.col) = st; }

    pub fn next_token(&mut self) -> Result<Token, ParseError> {
        self.skip();
        if self.eof() { return Ok(Token { typ: TokenType::Eof, value: String::new(), line: self.line, col: self.col }); }
        let (line, col, ch) = (self.line, self.col, self.cur());
        let one = |typ, v| Token { typ, value: v, line, col };
        match ch {
            '=' => { self.adv(); Ok(one(TokenType::Equal, "=".into())) }
            ',' => { self.adv(); Ok(one(TokenType::Comma, ",".into())) }
            '.' => { self.adv(); Ok(one(TokenType::Dot, ".".into())) }
            '[' => { self.adv(); Ok(one(TokenType::LBracket, "[".into())) }
            ']' => { self.adv(); Ok(one(TokenType::RBracket, "]".into())) }
            '"' => self.read_string(line, col),
            '-' if self.peek().is_ascii_digit() => Ok(self.read_number(line, col)),
            c if c.is_ascii_digit() => Ok(self.read_number(line, col)),
            c if c.is_ascii_alphabetic() || c == '_' => Ok(self.read_ident(line, col)),
            _ => Err(ParseError::new("unexpected character", line, col)),
        }
    }

    fn read_string(&mut self, line: usize, col: usize) -> Result<Token, ParseError> {
        self.adv();
        let mut out = String::new();
        while !self.eof() && self.cur() != '"' {
            if self.cur() == '\\' {
                self.adv();
                if self.eof() { return Err(ParseError::new("unterminated escape", line, col)); }
                out.push(match self.cur() {
                    '"' => '"', 'n' => '\n', 't' => '\t', '\\' => '\\',
                    _ => return Err(ParseError::new("invalid escape", self.line, self.col)),
                });
                self.adv();
            } else { out.push(self.cur()); self.adv(); }
        }
        if self.eof() { return Err(ParseError::new("unterminated string", line, col)); }
        self.adv();
        Ok(Token { typ: TokenType::String, value: out, line, col })
    }

    fn read_number(&mut self, line: usize, col: usize) -> Token {
        let mut out = String::new();
        if self.cur() == '-' { out.push('-'); self.adv(); }
        while !self.eof() && self.cur().is_ascii_digit() { out.push(self.cur()); self.adv(); }
        if !self.eof() && self.cur() == '.' && self.peek().is_ascii_digit() {
            out.push('.'); self.adv();
            while !self.eof() && self.cur().is_ascii_digit() { out.push(self.cur()); self.adv(); }
        }
        Token { typ: TokenType::Number, value: out, line, col }
    }

    fn read_ident(&mut self, line: usize, col: usize) -> Token {
        let mut out = String::new();
        while !self.eof() && (self.cur().is_ascii_alphanumeric() || self.cur() == '_') { out.push(self.cur()); self.adv(); }
        let typ = if out == "do" { TokenType::Do } else if out == "end" { TokenType::End } else { TokenType::Identifier };
        Token { typ, value: out, line, col }
    }

    fn skip(&mut self) {
        while !self.eof() {
            if self.cur().is_whitespace() { self.adv(); }
            else if self.cur() == '#' { while !self.eof() && self.cur() != '\n' { self.adv() } }
            else { break; }
        }
    }

    fn adv(&mut self) { if self.eof() { return; } if self.cur() == '\n' { self.line += 1; self.col = 1 } else { self.col += 1 } self.pos += 1; }
    fn cur(&self) -> char { self.input[self.pos] }
    fn peek(&self) -> char { self.input.get(self.pos + 1).copied().unwrap_or('\0') }
    fn eof(&self) -> bool { self.pos >= self.input.len() }
}
